Makefile.am

```diff
@@ -67,5 +68,9 @@ SUBDIRS = \
   tests \
   tools
 
+if PERFORM_UNIT_TESTS
+SUBDIRS += tests
```

Comment:
I think this is much simpler so also other parts can be written like this.

Reply:
Hi @metalefty - I've just had a test of this, and it's working but I think the reasons it's working are quite subtle and not entirely what we expected.

SUBDIRS still contains the `tests` directory anyway (see line 68 above your addition), so lines 71 and 72 are just adding tests to SUBDIRS for a second time (i.e. effectively a no-operation).

If line 68 is removed, SUBDIRS will only contain `tests` if we're running unit tests which is the intention of lines 71 and 72. However this has the problem that `make check` succeeds if `check` isn't installed which isn't what we want. However, if line 68 isn't removed, `make check` fails if `check` isn't installed.

So I took out PERFORM_UNIT_TESTS entirely and I think that's what we want.

The reason it's OK is a subtlety in `tests/common/Makefile.am` and `test/memtest/Makefile.am` which I hadn't realised. These makefiles use `check_PROGRAMS` to define the targets, and these are only built if running `make check` - they are not built for `make all`. So we don't need to cater for the unit tests being built or not built as automake already does it for us.

There's a further complication(!) in that `tests/memtest/Makefile.am` contains a `noinst_PROGRAMS` which is built by default. However, `memtest` doesn't use the `check` package so that's OK. I suggest this is removed so that `memtest` is only built for `make check`.

I hope that makes sense - I know I've learned a lot looking into this.

Here are the changes I made which I think give us what we want:-

```diff
diff --git a/Makefile.am b/Makefile.am
index 13795164..0f7b5d3e 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -68,9 +68,5 @@ SUBDIRS = \
   tests \
   tools
 
-if PERFORM_UNIT_TESTS
-SUBDIRS += tests
-endif
-
 distclean-local:
        -rm -f xrdp_configure_options.h
diff --git a/configure.ac b/configure.ac
index 00de8f5c..b513536a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -378,8 +378,6 @@ else
     AC_MSG_NOTICE([libcheck not found, unit tests will be skipped])
 fi
 
-AM_CONDITIONAL(PERFORM_UNIT_TESTS,
-               [test "x$perform_unit_tests" = "xyes"])
 # -- end perform unit tests
 
 AC_SUBST([moduledir], '${libdir}/xrdp')
diff --git a/tests/memtest/Makefile.am b/tests/memtest/Makefile.am
index 8901722f..890f2716 100644
--- a/tests/memtest/Makefile.am
+++ b/tests/memtest/Makefile.am
@@ -6,9 +6,6 @@ if XRDP_DEBUG
 AM_CPPFLAGS += -DXRDP_DEBUG
 endif
 
-noinst_PROGRAMS = \
-  memtest
-
 check_PROGRAMS = \
   memtest
 
```

Reply:
Thanks for the investigation. I confirmed the change causes `make check` failure if the `check` library is not installed and the ordinary `make install` succeeds without building unit tests.  

Then, is the `--enable-tests`  flag that enforces unit tests still necessary?

BTW, `--enable-tests`, "enforce unit tests", "skip unit tests" sound a little bit odd now. The first two enable strict dependency checking on `check` but not enforcing unit tests. The last one doesn't skip unit tests actually.  The flag does not change the behaviour during the build at all.

Reply:
I think the `--enable-tests` flags is still useful if only for `make distcheck`. In this case, if `check` isn't installed the configure fails with an informative message rather than just failing when it gets to `make check`.

So the meaning of the flag is more along the lines of 'ensure dependencies for the tests are installed' (I think).